Add blog post on translating R Markdown

r-ash opened the pull request:
Very keen for any feedback on my writing or how to word stuff differently/better

r-ash committed:
Add blog post on translating R Markdown

weshinsley commented on content/blog/translating-rmd.md:
`````diff
@@ -0,0 +1,132 @@
+:::
+````
+
+This syntax enables writing R Markdown as normal, no need for calls to `cat` or to manually manage formatting. It gives an easy way to add a second language to an existing document as you just need to surround existing code with `:::` and add the second language below. The downside is this will run every block of code and hide the sections for language not being used. Because of this it is not th ...
`````
third line - add *s* to language*s*

richfitz replied:
Could you add a link to `:::` in the Markdown docs and/or a short explanation (perhaps a footnote) explaining what it's doing? It seems quite mysterious

weshinsley commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+## Conclusion
+
+When translating R Markdown in production projects we've used a mix of approaches. Custom blocks provide a good way to translate large sections of text as they enable us to take advantage of automatic formatting and syntax highlighting. Combining with code chunks for translating slow to produce plots enables us to keep the generation of documents performant.
```
Perhaps hyphenate slow-to-produce ?

weshinsley commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+The problem with all of the methods above is they don't scale well. They work fine for 2 or 3 languages but if you want to provide translations for more than this then the document starts to get long and complicated. There is a more scalable approach detailed in [StatnMap blog post](https://statnmap.com/2017-10-06-translation-rmarkdown-documents-using-data-frame/) which uses a data frame of strin ...
+
+I think would be worth investigating a scalable approach via multiple R Markdown documents, one for each language. Making the name conform to some convetion `<report>_fr.Rmd`, `<report>_en.Rmd` etc. would then give a way to build input filename and pass to `rmarkdown::render` to generate the report in the correct language programatically. This would mean it would be straightforward to add a new l ...
```
insert I think [it] would be worth.. and typo on conve[n]tion

weshinsley reviewed (commented):
A couple of tiny things, but I think this is a great post, clearly written and described.

Some random thoughts, - not sure if I'm on the right lines or whether you've thought of these already - but a quick stream of consciousness:-

In your last para, were you thinking reports like report_fr.Rmd could be auto-generated? Eg - you could think about writing one R Markdown document where the "translatable" bits of text are marked up in some minimal way - perhaps that gets ignored on HTML rendering - (eg, <!--tran>Graph of random points<!--/tran> - then have a lookup table like the data-frame from the previous point that tells you the equivalent phrases in the different languages... and a search replace sort of function along the lines of rmarkdown::render(translate("report.Rmd"), "fr")

OR, if there's a lot of data processing, perhaps some sort of search/replace on the html output might be quicker than regenerating the report per language... (unless the important text is baked into PNGs by then...)

mwoodbri commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+## Code chunks
+
+The first method uses conditionally evaluated [code chunks](https://bookdown.org/yihui/rmarkdown/r-code.html). We can use R Markdown chunk options `include` and `eval` to hide ouput and stop a chunk from being evaluated. Combining with `results="asis"` this allows us to output text conditionally based on the `lang` parameter.
```
```suggestion
The first method uses conditionally evaluated [code chunks](https://bookdown.org/yihui/rmarkdown/r-code.html). We can use R Markdown chunk options `include` and `eval` to hide output and stop a chunk from being evaluated. Combining with `results="asis"` this allows us to output text conditionally based on the `lang` parameter.
```

mwoodbri commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+## Translating Rmd
+
+As part of our work on [naomi](../../projects/naomi) we've been asked to produce reports from R which can be distributed to country teams across Sub-Saharan Africa in English, French and Portuguese. R Markdown has no prescribed method for translating documents, but there are many options available to make this possible. This post covers 3 possible approaches: code chunks, custom blocks and tabset ...
```
Just in case anyone is familiar with R but not R Markdown then you could add a link:
```suggestion
As part of our work on [naomi](../../projects/naomi) we've been asked to produce reports from R which can be distributed to country teams across Sub-Saharan Africa in English, French and Portuguese. [R Markdown](https://rmarkdown.rstudio.com/docs/) has no prescribed method for translating documents, but there are many options available to make this possible. This post covers 3 possible approaches: code chunks, custom blocks and tabsets.
```

mwoodbri commented on content/blog/translating-rmd.md:
`````diff
@@ -0,0 +1,132 @@
+
+````r
+---
+title: '&nbsp;'
`````
I'm guessing this is just because the title is required? Even so, it's slightly cryptic - maybe a dummy title would avoid any confusion.

richfitz replied:
Agreed. Also can the title be translated with this approach?

mwoodbri reviewed (commented):
Well I learnt somthing! Looks good, and I like the GIF.

mwoodbri commented on content/blog/translating-rmd.md:
`````diff
@@ -0,0 +1,132 @@
+```
+````
+
+This has the advantage that the chunk will only be evaulated for the output language. This is particularly important if any of the plots we're generating take a long time to compute. The downside is the text needs be wrapped in calls to `cat` and the formatting of the text has to be done manually. For example we need to add the line break at the end of the text or the plot and text will appear on ...
`````
```suggestion
This has the advantage that the chunk will only be evaluated for the output language. This is particularly important if any of the plots we're generating take a long time to compute. The downside is the text needs be wrapped in calls to `cat` and the formatting of the text has to be done manually. For example we need to add the line break at the end of the text or the plot and text will appear on the same line. As everything is written as a code chunk this means RStudio syntax highlighting for the text is not available.
```

r-ash committed:
Update content/blog/translating-rmd.md

Co-authored-by: Mark Woodbridge <1724545+mwoodbri@users.noreply.github.com>

r-ash committed:
Update content/blog/translating-rmd.md

Co-authored-by: Mark Woodbridge <1724545+mwoodbri@users.noreply.github.com>

r-ash committed:
Update content/blog/translating-rmd.md

Co-authored-by: Mark Woodbridge <1724545+mwoodbri@users.noreply.github.com>

EmmaLRussell reviewed (approved):
This looks great to me. I like the way the final section leaves the door open for a sequel..

Just echoing what @weshinsley said, I found myself wondering how you could avoid replicating logic, both in the solutions you covered here and in the possible future approach you described (multiple markdowns). Maybe just a case of using shared functions or some R trick I don't know about - but in the next installment might be good to go into this, with some more real-world examples.

richfitz commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+ - translation
+---
+
+## Translating Rmd
```
I don' think you want this header, it will appear after hugo renders the title

richfitz commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+## Translating Rmd
+
+As part of our work on [naomi](../../projects/naomi) we've been asked to produce reports from R which can be distributed to country teams across Sub-Saharan Africa in English, French and Portuguese. [R Markdown](https://rmarkdown.rstudio.com/docs/) has no prescribed method for translating documents, but there are many options available to make this possible. This post covers 3 possible approaches ...
```
prescribed method -> built-in support

richfitz replied:
You might also foreshadow that the 3 approaches have different strengths/weaknesses and different levels of complexity to implement/maintain.

richfitz commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+## Code chunks
+
+The first method uses conditionally evaluated [code chunks](https://bookdown.org/yihui/rmarkdown/r-code.html). We can use R Markdown chunk options `include` and `eval` to hide output and stop a chunk from being evaluated. Combining with `results="asis"` this allows us to output text conditionally based on the `lang` parameter.
```
Perhaps add here what the consequences are? This will require n different renders for n different languages, with manually-curated duplication between chunks?

richfitz commented on content/blog/translating-rmd.md:
`````diff
@@ -0,0 +1,132 @@
+```
+````
+
+This has the advantage that the chunk will only be evaluated for the output language. This is particularly important if any of the plots we're generating take a long time to compute. The downside is the text needs be wrapped in calls to `cat` and the formatting of the text has to be done manually. For example we need to add the line break at the end of the text or the plot and text will appear on ...
`````
The duplication here looks like the biggest disadvantage to me, plus scaling this to (say) 5 languages would see very cumbersome

richfitz commented on content/blog/translating-rmd.md:
```diff
@@ -0,0 +1,132 @@
+
+The problem with all of the methods above is they don't scale well. They work fine for 2 or 3 languages but if you want to provide translations for more than this then the document starts to get long and complicated. There is a more scalable approach detailed in [StatnMap blog post](https://statnmap.com/2017-10-06-translation-rmarkdown-documents-using-data-frame/) which uses a data frame of strin ...
+
+I think would be worth investigating a scalable approach via multiple R Markdown documents, one for each language. Making the name conform to some convetion `<report>_fr.Rmd`, `<report>_en.Rmd` etc. would then give a way to build input filename and pass to `rmarkdown::render` to generate the report in the correct language programatically. This would mean it would be straightforward to add a new l ...
```
Sorry, I don't really follow the proposed solution here

richfitz reviewed (commented):
Did we ever investigate using traduire to make this work? Or is that the topic of a future blog post?

r-ash committed:
Add translations to document titles and fix issues after review

r-ash committed:
Rework summary paragraph

r-ash commented:
Thanks all for feedback. added translations of the titles and a discussion of that. Also reworked last paragraph to remove suggestion and keep it more open ended. Fixed up the typos too.

Next one can expand on this and maybe talk about using `traduire` to help remove some of the duplication.

r-ash committed:
Fix some typos and grammar issues

r-ash committed:
Refer to internationalisation over translation as that is actually what we are doing

richfitz reviewed (approved):
LGTM! :shipit:

mwoodbri reviewed (approved):
Very nice - especially most recent changes

EmmaLRussell reviewed (approved):
Looks good!

r-ash committed:
Update date
